Haskell-Generate/GenRustJets.hs

```diff
@@ -0,0 +1,377 @@
+         | otherwise = "()"
+    envArg | Just "Bitcoin" == moduleName mod = "_env"
+           | otherwise = "env"
+    envBody | Nothing == moduleName mod = "env"
+            | Just "Bitcoin" == moduleName mod = "unimplemented!(\"Unspecified CJetEnvironment for Bitcoin jets\")"
+            | otherwise = "env.c_tx_env()"
+
+  encodeStub = vsep
+    [ "fn encode<W: Write>(&self, w: &mut BitWriter<W>) -> std::io::Result<usize> {"
+    , "    todo!()"
+    , "}"
+    ]
+  decodeStub = vsep
+    [ "fn decode<I: Iterator<Item = u8>>(bits: &mut BitIter<I>) -> Result<Self, Error> {"
+    , "    todo!()"
+    , "}"
+    ]
+
+rustJetEnum :: Module -> Doc a
+rustJetEnum mod = vsep
+ [ pretty $ "/// " ++ rustModuleName mod ++ " jet family"
+ , "#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]"
+ , nest 4 (vsep (pretty ("pub enum " ++ rustModuleName mod ++ " {")
+   : [pretty (jetName jet) <> comma | (SomeArrow jet) <- moduleJets mod]))
+ , "}"
+ ]
+
+rustHeader :: Doc a
+rustHeader = "/* This file has been automatically generated. */"
+
+rustImports :: Module -> Doc a
+rustImports mod = vsep (map (<> semi)
+  ([ "use crate::bititer::BitIter"
+  , "use crate::bitwriter::BitWriter"
+  , "use crate::jet::type_name::TypeName"
+  , "use crate::jet::Jet"
+  , "use crate::merkle::cmr::Cmr"
+  , "use crate::{decode_bits, Error}"
+  , "use bitcoin_hashes::sha256::Midstate"
+  , "use simplicity_sys::CFrameItem"
+  , "use std::io::Write"
+  ] ++ envImports))
+ where
+  envImports | Nothing == moduleName mod = []
+             | Just "Bitcoin" == moduleName mod = ["use crate::jet::bitcoin::BitcoinEnv"]
+             | Just name <- moduleName mod =
+             [ pretty $ "use crate::jet::"++map toLower name++"::"++name++"Env"
+             , pretty $ "use simplicity_sys::C"++name++"TxEnv"
+             ]
+
+rustJetDoc :: Module -> SimpleDocStream a
+rustJetDoc mod = layoutPretty layoutOptions $ vsep (map (<> line)
+  [ rustHeader
+  , rustImports mod
+  , rustJetEnum mod
+  , rustJetImpl mod
+  ])
+
+rustFFIImports :: Doc a
+rustFFIImports = vsep (map (<> semi)
+  [ "use std::ffi::c_void"
+  , "use super::c_env::CElementsTxEnv"
+  , "use super::frame_ffi::CFrameItem"
+  ])
+
+rustFFISigs :: Module -> Doc a
+rustFFISigs mod = vsep
+  [ nest 4 (vsep ("extern \"C\" {": ((<> semi) . pretty . sig <$> moduleJets mod)))
+  , "}"
+  ]
+ where
+  sig (SomeArrow jet) = "pub fn "++cJetName jet++"(dst: *mut CFrameItem, src: CFrameItem, env: *const "++envType++") -> bool"
+   where
+    envType | CoreModule <- jetModule jet = "c_void"
+            | ElementsModule <- jetModule jet = "CElementsTxEnv"
+
+rustFFIDoc :: Module -> SimpleDocStream a
+rustFFIDoc mod = layoutPretty layoutOptions $ vsep (map (<> line)
+  [ rustHeader
+  , rustFFIImports
+  , rustFFISigs mod
+  ])
+
+rustWrapperImports :: Doc a
+rustWrapperImports = vsep (map (<> semi)
+  [ "use crate::CElementsTxEnv"
+  , "use super::{frame_ffi::CFrameItem, jets_ffi}"
```

Comment:
Promise this is the last functional thing. 
```diff
- "use super::{frame_ffi::CFrameItem, jets_ffi}"
+ "use super::{frame_ffi::CFrameItem, elements_ffi}"
```

Reply:
Done.